dali/operators/image/crop/bbox_crop.cc

```diff
@@ -14,143 +14,734 @@
+                     make_string("`shape_layout` should be a permutation of ", default_shape_layout,
+                                 "` for the provided inputs"));
+        auto perm = GetDimIndices(shape_layout_, default_shape_layout);
+        for (int sample = 0; sample < static_cast<int>(input_shape_.size()); sample++) {
+          TensorShape<> in_shape = input_shape_[sample];
+          TensorShape<> crop_shape = crop_shape_[sample];
+
+          DALI_ENFORCE(crop_shape.sample_dim() == ndim,
+                    make_string("Unexpected number of dimensions. Expected ", ndim, ", got ",
+                                crop_shape.sample_dim()));
+
+          for (int i = 0; i < ndim; i++) {
+            int axis = perm[i];
+            DALI_ENFORCE(
+                crop_shape[axis] <= in_shape[axis],
+                make_string("Crop shape can't exceed input shape dimensions. Got crop_shape=",
+                            crop_shape, ", input_shape=", in_shape));
+            input_shape_[sample][i] = in_shape[axis];
+            crop_shape_[sample][i]  = crop_shape[axis];
+          }
+        }
+      }
+    }
+    return false;
   }
-}
 
-template <>
-void RandomBBoxCrop<CPUBackend>::RunImpl(SampleWorkspace &ws) {
-  const auto &boxes_tensor = ws.Input<CPUBackend>(0);
+  void RunImpl(SampleWorkspace &ws) override {
+    const auto &boxes_tensor = ws.Input<CPUBackend>(0);
+    auto nboxes = boxes_tensor.dim(0);
+    auto ncoords = ndim * 2;
+
+    std::vector<Box<ndim, float>> bounding_boxes(nboxes);
+    ReadBoxes(make_span(bounding_boxes),
+              make_cspan(boxes_tensor.data<float>(), boxes_tensor.size()), bbox_layout_);
+
+    std::vector<int> labels;
+    if (has_labels_) {
+      const auto &labels_tensor = ws.Input<CPUBackend>(1);
+      auto nlabels = labels_tensor.dim(0);
+      DALI_ENFORCE(nlabels == nboxes,
+        make_string("Unexpected number of labels. Expected: ", nboxes, ", got ", nlabels));
+      labels.resize(nlabels);
+      const auto *label_data = labels_tensor.data<int>();
+      for (int i = 0; i < nlabels; i++) {
+        labels[i] = label_data[i];
+      }
+    }
+
+    int sample = ws.data_idx();
+    ProspectiveCrop prospective_crop;
+    while (!prospective_crop.success) {
+      prospective_crop =
+          FindProspectiveCrop(make_cspan(bounding_boxes), make_cspan(labels), sample);
+    }
+
+    WriteCropToOutput(ws, prospective_crop.crop);
+    WriteBoxesToOutput(ws, make_cspan(prospective_crop.boxes));
+
+    if (has_labels_) {
+      DALI_ENFORCE(
+          prospective_crop.boxes.size() == prospective_crop.labels.size(),
+          make_string("Expected boxes.size() == labels.size(). Received: ",
+                      prospective_crop.boxes.size(), " != ", prospective_crop.labels.size()));
+      WriteLabelsToOutput(ws, make_cspan(prospective_crop.labels));
+    }
+  }
+
+ private:
+  struct ProspectiveCrop {
+    bool success = false;
```

Comment:
What does `success` should denote in this context? I don't see it being updated after creating, so if it's about telling that the ProspeciveCrop is actually the crop we want to have, why not assume that every `ProspectiveCrop` will be the crop we want to have?

Reply:
`ProspectiveCrop` can either be a prospective crop or not a prospective crop (when you couldn't find it). This flag is just used to see if we could find a prospective crop or we should try again.
I haven't added this type, just moved it around